### PR TITLE
Basic implementation of WeakMap/WeakSet

### PR DESCRIPTION
**Pull request by arv**
Uses an expando on the key

Issue #898

### CODE REVIEW

**Comment by arv** [comment]
@johnjbarton PTAL

The motivation for finally doing this is that I want to get rid of our `createPrivateName` and replace all of those with weak maps.

**Comment by arv** [comment]
This does not handle non extensible/frozen objects but we can update to support that and or just switch to core-js (see issue #1976).

File: src/runtime/polyfills/WeakMap.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+const sentinel = {};
+
+export class WeakMap {
+  constructor() {
+    this.name_ = newUniqueString();
+  }
+
+  set(key, value) {
+    if (!isObject(key)) throw new $TypeError('key must be an object');
```
Comment (johnjbarton) [inline src/runtime/polyfills/WeakMap.js]: I wonder how these error message will appear? Will it be clear or should it start with "WeakMap key must be an object"?

File: src/runtime/polyfills/WeakMap.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  let {WeakMap, Symbol} = global;
+  if (!WeakMap || !hasNativeSymbol()) {
+    return true;
+  }
+  try {
+    let o = {};
+    let wm = new WeakMap([[o, false]]);
+    return wm.get(o);
+  } catch (e) {
+    return false;
```
Comment (johnjbarton) [inline src/runtime/polyfills/WeakMap.js]: Is is obvious why throw results in `needsPolyfill()` false?

File: src/runtime/polyfills/WeakSet.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+const $TypeError = TypeError;
+const {hasOwnProperty} = Object.prototype;
+
+export class WeakSet {
+  constructor() {
+    this.name_ = newUniqueString();
+  }
+
+  add(key) {
+    if (!isObject(key)) throw new $TypeError('value must be an object');
```
Comment (johnjbarton) [inline src/runtime/polyfills/WeakSet.js]: From the self-document point of view, `add(value)` would be clearer.

File: src/runtime/runtime.js [hunk 4]
```diff
@@ -56,14 +57,14 @@
...
-  var counter = 0;
+  var counter = Date.now() % 1e9;
 
   /**
    * Generates a new unique string.
    * @return {string}
    */
   function newUniqueString() {
-    return '__$' + Math.floor(Math.random() * 1e9) + '$' + ++counter + '$__';
+    return '__$' + (random() * 1e9 >>> 1) + '$' + ++counter + '$__';
```
Comment (johnjbarton) [inline src/runtime/runtime.js]: A reference to the analysis for this construct or an analysis in a comment would help clarify the design.

**Comment by johnjbarton** [comment]
LGTM, though using eg corejs is attractive from the maintenance viewpoint.

**Comment by arv** [comment]
I realized this was still pending. I added support for frozen keys (which have to be strongly held) as well as fixed few issues regarding non object keys.

PTAL.

**Comment by arv** [comment]
TODO: Handle the case where a key gets added and then later gets frozen...

**Comment by arv** [comment]
I thought more about the case where a key is later frozen. It gets complicated and I don't think it is worth supporting at this point.

**Pull request closed without merging**